File: python/p022.py
#!/usr/bin/python

# Using names.txt (right click and 'Save Link/Target As...'), a 46K text file containing over five-thousand first
# names,  begin by sorting it into alphabetical order. Then working out the alphabetical value for each name,
# multiply this value by its alphabetical position in the list to obtain a name score.

# For example, when the list is sorted into alphabetical order, COLIN, which is worth 3 + 15 + 12 + 9 + 14 = 53, is
# the 938th name in the list. So, COLIN would obtain a score of 938 × 53 = 49714.

# What is the total of all the name scores in the file?


def word_score(word):
    return sum([ord(char) - 96 for char in word.lower()])

words = []
with open("../data/p022.txt") as f:
    words = f.read().strip(' \t\n\r"').split('","')

total = 0
for i, word in enumerate(sorted(words), 1):
    total += i * word_score(word)

print(total)

File: rust/src/bin/p022.rs

use std::io::Read;
use std::fs::File;

fn word_score(word: &str) -> usize {
    word.bytes().map(|x: u8| x - 64).fold(0, |acc, x| acc + (x as usize))
}

fn main() {

    let mut total = 0;

    let mut f = File::open("../data/p022.txt").unwrap();
    let mut s = String::new();
    let _ = f.read_to_string(&mut s);

    let chars_to_trim: &[char] = &['\t', '\n', '\r', '"'];
    let mut words = s.trim_matches(chars_to_trim).split("\",\"").collect::<Vec<&str>>();
    words.sort();

    for (i, word) in words.iter().enumerate() {
        total += (i + 1) * word_score(word);
    }

    println!("{}", total);

}
